bls12_381/src/g1.rs

```diff
@@ -793,6 +831,383 @@ impl G1Projective {
+        &self.0
+    }
+}
+
+impl AsMut<[u8]> for G1Uncompressed {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0
+    }
+}
+
+impl Group for G1Projective {
+    type Scalar = Scalar;
+
+    fn random<R: RngCore + ?Sized>(rng: &mut R) -> Self {
+        loop {
+            let x = Fp::random(rng);
+            let greatest = rng.next_u32() % 2 != 0;
+
+            // Obtain the corresponding y-coordinate given x as y = sqrt(x^3 + 4)
+            let p = ((x.square() * x) + B).sqrt().map(|y| {
+                G1Affine {
+                    x,
+                    // Switch to the correct y-coordinate if necessary.
+                    y: if bool::from(y.lexicographically_largest()) ^ greatest {
+                        -y
+                    } else {
+                        y
+                    },
+                    infinity: 0.into(),
+                }
+            });
+
+            if p.is_some().into() {
+                let p = p.unwrap().to_curve();
+
+                if bool::from(!p.is_identity()) {
+                    return p;
+                }
+            }
+        }
+    }
+
+    fn identity() -> Self {
+        Self::identity()
+    }
+
+    fn generator() -> Self {
+        Self::generator()
+    }
+
+    fn is_identity(&self) -> Choice {
+        self.is_identity()
+    }
+
+    #[must_use]
+    fn double(&self) -> Self {
+        self.double()
+    }
+}
+
+impl Group for G1Subgroup {
+    type Scalar = Scalar;
+
+    fn random<R: RngCore + ?Sized>(rng: &mut R) -> Self {
+        loop {
+            let p = G1Projective::random(rng).mul_by_cofactor();
+
+            if bool::from(!p.is_identity()) {
+                return p;
+            }
+        }
+    }
+
+    fn identity() -> Self {
+        G1Subgroup(G1Projective::identity())
+    }
+
+    fn generator() -> Self {
+        G1Subgroup(G1Projective::generator())
+    }
+
+    fn is_identity(&self) -> Choice {
+        self.0.is_identity()
+    }
+
+    #[must_use]
+    fn double(&self) -> Self {
+        G1Subgroup(self.0.double())
+    }
+}
+
+impl WnafGroup for G1Projective {
+    fn recommended_wnaf_for_scalar(_scalar: &Self::Scalar) -> usize {
+        let num_bits = Self::Scalar::NUM_BITS as usize;
+
+        if num_bits >= 130 {
+            4
+        } else if num_bits >= 34 {
+            3
+        } else {
+            2
+        }
+    }
+
+    fn recommended_wnaf_for_num_scalars(num_scalars: usize) -> usize {
+        const RECOMMENDATIONS: [usize; 12] =
+            [1, 3, 7, 20, 43, 120, 273, 563, 1630, 3128, 7933, 62569];
+
+        let mut ret = 4;
+        for r in &RECOMMENDATIONS {
+            if num_scalars > *r {
+                ret += 1;
+            } else {
+                break;
+            }
+        }
+
+        ret
+    }
+}
+
+impl PrimeGroup for G1Subgroup {}
+
+impl CofactorGroup for G1Projective {
+    type Subgroup = G1Subgroup;
+
+    fn mul_by_cofactor(&self) -> Self::Subgroup {
```

Comment:
In discussion with @str4d:

We agreed to change this to `clear_cofactor()` (and use the implementation already available in `BLS12-381`).

```
    /// This function clears the cofactor by multiplying
    /// the group element by a multiple of the cofactor (not necessarily
    /// the actual cofactor) which effectively maps this group element
    /// to the prime order subgroup. This function is deterministic.
```